src/java/org/apache/cassandra/index/sai/disk/v2/sortedterms/SortedTermsReader.java

```diff
@@ -0,0 +1,305 @@
+
+        DirectMonotonicReader.Meta offsetsMeta =
+                DirectMonotonicReader.loadMeta(offsestMetaInput, meta.offsetBlockCount, DIRECT_MONOTONIC_BLOCK_SHIFT);
+        RandomAccessInput offsetSlice = termsDataBlockOffsets.randomAccessSlice(0, termsDataBlockOffsets.length());
+        this.blockOffsets = DirectMonotonicReader.getInstance(offsetsMeta, offsetSlice);
+    }
+
+    /**
+     * Returns the point id (ordinal) of the target term or the next greater if no exact match found.
+     * If reached the end of the terms file, returns <code>Long.MAX_VALUE</code>.
+     * Complexity of this operation is O(log n).
+     *
+     * @param term target term to lookup
+     */
+    public long getPointId(@Nonnull ByteComparable term)
+    {
+        Preconditions.checkNotNull(term, "term null");
+
+        try (TrieRangeIterator reader = new TrieRangeIterator(termsTrie.instantiateRebufferer(),
+                                                              meta.trieFP,
+                                                              term,
+                                                              null,
+                                                              true,
+                                                              true))
+        {
+            final Iterator<Pair<ByteSource, Long>> iterator = reader.iterator();
+            return iterator.hasNext() ? iterator.next().right : Long.MAX_VALUE;
+        }
+    }
+
+    /**
+     * Returns the total number of terms.
+     */
+    public long count()
+    {
+        return meta.count;
+    }
+
+    /**
+     * Opens a cursor over the terms stored in the terms file.
+     * <p>
+     * This does not read any data yet.
+     * The cursor is initially positioned before the first item.
+     * <p>
+     * The cursor is to be used in a single thread.
+     * The cursor is valid as long this object hasn't been closed.
+     * You must close the cursor when you no longer need it.
+     */
+    public @Nonnull Cursor openCursor() throws IOException
+    {
+        return new Cursor(termsData);
+    }
+
+    /**
+     * Allows reading the terms from the terms file.
+     * Can quickly seek to a random term by <code>pointId</code>.
+     * <p>
+     * This object is stateful and not thread safe.
+     * It maintains a position to the current term as well as a buffer that can hold one term.
+     */
+    @NotThreadSafe
+    public class Cursor implements AutoCloseable
+    {
+        private final IndexInput termsData;
+
+        // The term the cursor currently points to. Initially empty.
+        private final BytesRef currentTerm;
+
+        // The point id the cursor currently points to. -1 means before the first item.
+        private long pointId = -1;
+
+        Cursor(IndexInput termsData) throws IOException
+        {
+            this.termsData = termsData.clone();  // clone needed to get our own file pointer independently of other Cursors
```

Comment:
I am unaware of `IndexInput#clone` working in SAI.

Reply:
Will be improved in the next commit, see conversation in slack, also for other reasons.

Reply:
`clone` does not work in SAI.

Reply:
As have said it will be addressed (and won't use `clone`).
And BTW: if `clone` doesn't work in SAI, it is a bug.

Reply:
Open a jira